File: 587. Erect the Fence/Solution.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int[][] outerTrees(int[][] trees) {
        if (trees.length < 4)
            return trees;
        int mostLeftTreeIndex = 0;
        for (int i = 1; i < trees.length; i++) {
            if (trees[mostLeftTreeIndex][0] < trees[i][0])
                mostLeftTreeIndex = i;
        }
        Set<int[]> resultSet = new HashSet<>();
        resultSet.add(trees[mostLeftTreeIndex]);
        int current, candidate;
        current = mostLeftTreeIndex;
        while (true) {
            candidate = (current + 1) % trees.length;
            for (int i = 0; i < trees.length; i++) {
                if (isMoreClockWise(trees[current], trees[candidate], trees[i]) > 0) {
                    candidate = i;
                }
            }
            for (int[] is : trees) {
                if (isMoreClockWise(trees[current], trees[candidate], is) == 0)
                    resultSet.add(is);
            }
            current = candidate;
            if (current == mostLeftTreeIndex)
                break;
        }
        return resultSet.toArray(new int[][] {});
    }

    private int isMoreClockWise(int[] firstPoint, int[] secondPoint, int[] thirdPoint) {

        return ((thirdPoint[0] - firstPoint[0]) * (secondPoint[1] - firstPoint[1]))
                - ((secondPoint[0] - firstPoint[0]) * (thirdPoint[1] - firstPoint[1]));
    }
}

File: 587. Erect the Fence/Solution.py
from typing import List


class Solution:
    def outerTrees(self, trees: List[List[int]]) -> List[List[int]]:
        if (len(trees) < 4):
            return trees
        mostLeftTreeIndex: int = 0
        for i in range(len(trees)):
            if trees[i][0] > trees[mostLeftTreeIndex][0]:
                mostLeftTreeIndex = i
        currentTreeIndex = mostLeftTreeIndex
        result = set()
        result.add(tuple(trees[mostLeftTreeIndex]))

        def isMoreClockWise(first_tree: List[int], second_tree: List[int], third_tree: List[int],) -> int:
            return ((third_tree[0]-first_tree[0])*(second_tree[1]-(first_tree[1])))-((second_tree[0]-first_tree[0])*(third_tree[1]-first_tree[1]))

        while True:
            candidateIndex = (currentTreeIndex+1) % len(trees)
            for i, tree in enumerate(trees):
                if (isMoreClockWise(trees[currentTreeIndex], trees[candidateIndex], tree) > 0):
                    candidateIndex = i
            for tree in trees:
                if (isMoreClockWise(trees[currentTreeIndex], trees[candidateIndex], tree) == 0):
                    result.add(tuple(tree))

            currentTreeIndex = candidateIndex
            if (currentTreeIndex == mostLeftTreeIndex):
                return result
